Abstract Swift and C++ handleMessage differences

adetaylor opened the pull request:
#### e935d39eecd21246bfbf89a82e02cb0de291d8eb
<pre>
Abstract Swift and C++ handleMessage differences
<a href="https://bugs.webkit.org/show_bug.cgi?id=304185">https://bugs.webkit.org/show_bug.cgi?id=304185</a>
<a href="https://rdar.apple.com/165213339">rdar://165213339</a>

Reviewed by Geoffrey Garen.

This simplifies the &apos;handleMessage&apos; functions by removing a layer of nested
&apos;if&apos;s, which were introduced recently when we added the ability to call
directly into Swift message handling functions.

With this change, we move this new complexity into the existing
MethodSignatureValidation struct, allowing us to take the actual
message handlers back to their former simplicity.

There are no functional changes here.

Canonical link: <a href="https://commits.webkit.org/304670@main">https://commits.webkit.org/304670@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/122b42247adb55b2666c0bade89939a007a1ec58

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136268 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8625 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47548 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143979 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/160adf74-8ea4-4bdd-bdc8-dec657c123af) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9303 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8469 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104190 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/deb5989a-427b-4d83-ac1c-0264cff3a27f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139213 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6768 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122104 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85022 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/73e981c2-6384-4d7f-ad92-95e4870e1607) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6422 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4080 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4571 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115712 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146723 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8307 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40873 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112529 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8324 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6978 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112872 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28642 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6349 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118409 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8355 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36466 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8073 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71914 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8295 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8147 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebKit/Platform/IPC/HandleMessage.h:
```diff
@@ -361,23 +367,38 @@ struct MethodSignatureValidationImpl<std::tuple<MessageArgumentTypes...>, std::t
-    static constexpr bool completionHandlerIsWrapped = true;
+    using CompletionHandlerType = CompletionHandler<void(CompletionHandlerArgumentTypes...)>;
+    template<typename CH>
+    static Ref<WTF::RefCountable<WTF::CompletionHandler<CH>>> wrapCompletionHandler(CompletionHandler<CH>&& handler)
```
Not new in this patch, but it's curious that we need to wrap in `RefCountable` before calling into Swift. Swift supports move-only types via `borrowing` and `consuming`. Did we try that?

adetaylor replied:
Not yet supported by reverse interop, sadly - rdar://162361370

geoffreygaren commented on Source/WebKit/Platform/IPC/HandleMessage.h:
```diff
@@ -361,23 +367,38 @@ struct MethodSignatureValidationImpl<std::tuple<MessageArgumentTypes...>, std::t
     using CompletionHandlerType = CompletionHandler<void(CompletionHandlerArgumentTypes...)>;
-    static constexpr bool completionHandlerIsWrapped = false;
+    static constexpr std::identity wrapCompletionHandler { };
+    static constexpr std::identity getCompletionHandlerArgument { };
```
Let's call this `unwrapCompletionHandler`. `get` is not WebKit style, and `Argument` is not the behavior even though it's typical usage.

Wrap and unwrap are not perfect inverses here, but they're close enough.

adetaylor replied:
Will do, thanks

adetaylor committed:
Abstract Swift and C++ handleMessage differences
https://bugs.webkit.org/show_bug.cgi?id=304185
rdar://165213339

Reviewed by Geoffrey Garen.

This simplifies the 'handleMessage' functions by removing a layer of nested
'if's, which were introduced recently when we added the ability to call
directly into Swift message handling functions.

With this change, we move this new complexity into the existing
MethodSignatureValidation struct, allowing us to take the actual
message handlers back to their former simplicity.

There are no functional changes here.

Canonical link: https://commits.webkit.org/304670@main
